### PR TITLE
Replacing Webpack(er) with Vite (SCP-4121)

### PR DESCRIPTION
**Pull request by devonbush**
SCP-4121.  This is ready for review, as it works well for me, Jon, and Jean (and there's a lot here, so it will be good to get started soon).  We should hold off on merging it until we can figure out what's going on with Emily's environment -- this doesn't build at all for her.  

The two concepts I think that are most important to learn:
1. In development mode, our Vite/React javascript will be loaded and execute *after* any inline scripts in our rails templates.  (the opposite will be true in production).  This required a number of changes, such as the new `renderComponent()` methods in both application.js files.  
2. Hot module reload operates on components.  Non-component exports from a component file (e.g. if a helper function or constant is exported too) will often force a page reload if anything in that file changes.  So I've started migrating some non-component exports to instead be set on the components themselves.  

A document outlining more detail is [here](https://docs.google.com/document/d/171bD5dgP1gIL5gMnQXorn-w6wCPvXKemxKqXlsMZDoc/edit#)

TO TEST:

1. stop any existing servers
2. checkout branch db-vite-build
4. run bundle install
5. run bundle exec vite install
6. run yarn install
7. run rails s in a terminal
8. run bin/vite dev in another terminal
9. go to https://localhost:3000/ and confirm homepage loads
10. click around to various parts of the application, ensuring functionality works.  Then make changes to some javascript files (e.g. StudySearchResult.jsx) and confirm live reload works.  There may be many files which still trigger a full page reload unnecessarily (see learning 2 above), but we can address these as we work on those files.

### CODE REVIEW

**Review by bistline — APPROVED** [review]
Given the size of the PR, I haven't gone through each file.  But I did confirm the functionality.  It would be good to check for debug `console.log()` statements, as I saw one or two.  Other than that, this works, and is much faster!

File: app/javascript/components/visualization/controls/ScatterPlotLegend.jsx [hunk 1]
```diff
@@ -127,7 +126,7 @@ function LegendEntry({
         </div>
         <div className="scatter-legend-entry">
           <span className="legend-label" title={entry}>{entry}</span>
-          <span className="num-points" title={`${numPoints} points in this group`}>{numPoints}</span>
+          <span className="num-points" title={`${numPoints} points in this crazy group`}>{numPoints}</span>
```
Comment (ehanna4) [inline app/javascript/components/visualization/controls/ScatterPlotLegend.jsx:129]: is the word 'crazy' intended to be here?

File: config/initializers/content_security_policy.rb [hunk 2]
```diff
@@ -10,9 +10,21 @@
...
+#    policy.script_src *policy.script_src, :unsafe_eval, "http://#{ ViteRuby.config.host_with_port }" if Rails.env.development?
+
+    # You may need to enable this in production as well depending on your setup.
+#    policy.script_src *policy.script_src, :blob if Rails.env.test?
+
 #   policy.style_src   :self, :https
+    # Allow @vite/client to hot reload style changes in development
+#    policy.style_src *policy.style_src, :unsafe_inline if Rails.env.development?
+
 #   # If you are using webpack-dev-server then specify webpack-dev-server host
```
Comment (ehanna4) [inline config/initializers/content_security_policy.rb:23]: Should this be updated to remove the 'webpack-dev-server' specific text?

**Review by ehanna4 — APPROVED** [review]
@jlchang and I reviewed the files and anything funky we saw has now been fixed and i am able to successfully run this locally

**Comment by devonbush** [comment]
Build failure is false positive due to test running too close to midnight

**Pull request merged**